v-next/hardhat-keystore/src/internal/ui/user-interactions.ts

```diff
@@ -0,0 +1,120 @@
+    await this.#displayMessage("");
+
+    let password: string | undefined;
+
+    while (password === undefined) {
+      password = await this.#requestSecretInput(`Enter your password: `);
+
+      if (!PASSWORD_REGEX.test(password)) {
+        password = undefined;
+        await this.#displayMessage(chalk.red("Invalid password!"));
+      }
+    }
+
+    let confirmPassword: string | undefined;
+    while (confirmPassword === undefined) {
+      confirmPassword = await this.#requestSecretInput(
+        "Please confirm your password: ",
+      );
+
+      if (password !== confirmPassword) {
+        await this.#displayMessage(chalk.red("Passwords do not match!"));
+        confirmPassword = undefined;
+      }
+    }
+
+    return confirmPassword;
+  }
+
+  public async requestSecretFromUser(): Promise<string> {
+    return this.#requestSecretInput("Enter secret to store: ");
+  }
+
+  public async displayInvalidKeyErrorMessage(key: string): Promise<void> {
+    await this.#displayMessage(
+      chalk.red(
+        `Invalid value for key: "${key}". Keys can only have alphanumeric characters and underscores, and they cannot start with a number.`,
+      ),
+    );
+  }
+
+  public async displayKeyNotFoundErrorMessage(key: string): Promise<void> {
+    await this.#displayMessage(chalk.red(`Key "${key}" not found`));
+  }
+
+  public async displayNoKeystoreSetErrorMessage(): Promise<void> {
+    await this.#displayMessage(
+      `No keystore found. Please set one up using ${chalk.blue.italic("npx hardhat keystore set {key}")} `,
+    );
+  }
+
+  public async displaySecretCannotBeEmptyErrorMessage(): Promise<void> {
+    await this.#displayMessage(chalk.red("The value cannot be empty."));
+  }
+
+  public async displayKeyAlreadyExistsWarning(key: string): Promise<void> {
+    await this.#displayMessage(
+      chalk.yellow(
+        `The key "${key}" already exists. Use the ${chalk.blue.italic("--force")} flag to overwrite it.`,
+      ),
+    );
+  }
+
+  public async displayKeyListInfoMessage(keys: string[]): Promise<void> {
+    await this.#displayMessage("Keys:");
+    for (const key of keys) {
+      await this.#displayMessage(key);
+    }
+  }
+
+  public async displayKeyRemovedInfoMessage(key: string): Promise<void> {
+    await this.#displayMessage(`Key "${key}" deleted`);
+  }
+
+  public async displayKeySetInfoMessage(key: string): Promise<void> {
+    await this.#displayMessage(`Key "${key}" set`);
+  }
+
+  public async displayNoKeysInfoMessage(): Promise<void> {
+    await this.#displayMessage("The keystore does not contain any keys.");
+  }
+
+  public async displayValueInfoMessage(value: string): Promise<void> {
+    await this.#displayMessage(value);
+  }
+
+  async #displayMessage(message: string): Promise<void> {
+    await this.#userInterruptions.displayMessage(PLUGIN_ID, message);
+  }
+
+  async #requestSecretInput(inputDescription: string): Promise<string> {
+    return this.#userInterruptions.requestSecretInput(
```

Comment:
Instead of holding a reference to the user interruption manager I think this could just return the messages, so that it's approximately equally testable, and you don't have to couple this.

Note that if you do want to test that the interruption gets that message, you can set a dynamic hook handler from the test, assert the message, and then remove it.

Reply:
Based on the UserInteractions and interruptions comments I made the following moves:

* Changed user interactions into `UserDisplayMessages` where each static method just returns the string of the method
* In `get`, `list`, `delete`, `set` use poor man's dependency injection to inject a `consoleLog` defaulting to `console.log` allowing the tests to be ported
* Remove the interruptions class that is no longer needed.


[`f0efbb0` (#5654)](https://github.com/NomicFoundation/hardhat/pull/5654/commits/f0efbb0df60a0317da4f5bf4abf88682c6474ece)